src/utils/narrow.js

```diff
@@ -304,6 +304,55 @@ export const canSendToNarrow = (narrow: Narrow): boolean =>
     search: () => false,
   });
 
+/**
+ * Answers the question, "What narrows do this message appear in?"
+ *
+ * The message's flags must be in `flags`; `message.flags` is ignored.  This
+ * makes it the caller's responsibility to deal with the ambiguity in our
+ * Message type of whether the message's flags live in a `flags` property or
+ * somewhere else.
+ */
+export const getNarrowsForMessage = (
+  message: Message | Outbox,
+  ownEmail: string,
+  flags: $ReadOnlyArray<string>,
+): Narrow[] => {
+  const result: Narrow[] = [];
+
+  // All messages are in the home narrow.
+  result.push(HOME_NARROW);
+
+  // A message is either a private message (and `.display_recipient`
+  // is an array) or a stream message (and `.display_recipient` is the
+  // stream name, a string).
+  if (Array.isArray(message.display_recipient)) {
+    result.push(ALL_PRIVATE_NARROW);
+    if (message.display_recipient.length > 1) {
+      result.push(
+        groupNarrow(message.display_recipient.filter(x => x.email !== ownEmail).map(x => x.email)),
+      );
+    } else if (message.display_recipient.length === 1) {
+      result.push(privateNarrow(message.display_recipient[0].email));
+    } else {
+      // FIXME(crunchy): This shouldn't happen; reject at the edge.
+      result.push(groupNarrow([]));
+    }
+  } else {
+    result.push(topicNarrow(message.display_recipient, message.subject));
+    result.push(streamNarrow(message.display_recipient));
+  }
+
+  if (flags.includes('mentioned') || flags.includes('wildcard_mentioned')) {
+    result.push(MENTIONED_NARROW);
+  }
+
+  if (flags.includes('starred')) {
+    result.push(STARRED_NARROW);
+  }
+
```

Comment:
Let's add a comment mentioning search narrows, saying they're excluded.

In fact, that should probably be in the jsdoc -- just like `isMessageInNarrow`.